### ISSUE TITLE
🎛️ Migrate `initSlugAutoPopulate` to a Stimulus Controller `w-sync`

### ISSUE THREAD

**Issue by lb-** [wagtail/wagtail#10088]
> ℹ️ **Part of the [Stimulus  🎛️ RFC 78](https://github.com/wagtail/rfcs/pull/78)**

### Is your proposal related to a problem?

There is a core JavaScript implementation that allows for syncing the slug field with the title field as the user updates the title field. This implementation is quite rigid and it would be good to move this to a Stimulus controller, giving us an ability to provide more flexibility and have this functionality unit tested.

### Describe the solution you'd like

Implementation of a `w-sync` controller that allows the controlled element to 'sync' its value to some other input. It should allow for one or more elements to be selected via a selector string and then those elements become the target. Additionally, it should have the ability to turn on/off this sync behaviour based on a check.

#### Example usage

Given the HTML below, the 'title' field will be able to sync its value with the 'slug' field in the following ways;

* When the title field is focused, it will perform a check against the value of the slug field, if they are **not** the same it will enter the 'inactive' mode (a value) that means further apply/clear methods will do nothing. This is optional but allows for a value to sync only when they are not made different by user interaction.
* Assuming the values are different on 'check', when the title field changes, it will apply its value to the slug field's value.
* Assuming the values are different on 'check', when the title field receives the 'cut' event, it will completely clear the value on the slug field.
* Finally, the apply/clear methods can be debounced by the delay value (500ms), this value will default to 0 seconds

```html
<div>
  <input
    type="text"
    name="title"
    value="Anadama"
    placeholder="Page title*"
    maxlength="255"
    id="id_title"
    data-controller="w-sync"
    data-action="focus->w-sync#check change->w-sync#apply cut->w-sync#clear"
    data-w-sync-delay-value="500"
    data-w-sync-target-value="body:not(.page-is-live) #id_slug"
  />
  <input id="id_slug" name="slug" type="text" />
</div>

```

#### Part 1 - Create a `w-sync` Stimulus controller

This controller should be written in TypeScript and have a baseline of unit tests, it should have the following specification

* The controller should have the type of `HTMLInputElement`
* value: `inactive` or `inert` or maybe just `disabled`, boolean, default is false (see check method for usage)
* value: `delay`, used for the debounced wrapper around `clear` & `apply` methods.
* value: `target`, a string, no default or empty string (ok to error if not set), expects a CSS style selector
* method: `connect` - dispatches an event `start` on each of the target elements, if preventDefault is called, the controller will disconnect itself. Event dispatch must not bubble.
* method: `getTargetElements` - used as a convenience wrapper around the selection of a target elements from the target value. Very important that this uses `document.querySelectorAll` so that multiple targets are supported. This also makes the code simpler for typing.
* method: `check` - dispatches an event `check` on each of the target elements, if ANY of the events get their default prevented, then the controller goes into inactive mode (`this.disabledValue = true`). Must dispatch with a reference to the `this.element` in the event detail (e.g. `const event = this.dispatchEvent('check', { bubbles: false, cancellable: true, detail: { element: this.element, value: this.element.value } });`.
* method: `apply` - first dispatches an event, similar to `check` but with the name `apply` and if default is prevented does nothing else, if not it will update the target elements' values
* method: `clear` - similar to apply, but instead of applying a value, it just resets it to an empty string.
* method: `disconnect` - should clear any debounced timers and also dispatch an event `stop` to the target elements.

**Start small** Build this in small steps, driven by unit test writing, the `clear` and also the start/stop events can be done last as they are not critical to the core functionality and are nice to have.

#### Part 2 - Adopt the sync field in the Page title form field

In `wagtail/admin/panels/page_utils.py` add the relevant data attributes, example below.

```python
def set_default_page_edit_handlers(cls):

    cls.content_panels = [
        FieldPanel(
            "title",
            classname="title",
            widget=forms.TextInput(
                attrs={
                    "placeholder": format_lazy(
                        "{title}*", title=gettext_lazy("Page title")
                    ),
                    "data-controller": "w-sync",
                    "data-w-sync-target-value": "body:not(.page-is-live) #id_slug",
                    "data-action": "focus->w-sync#check blur->w-sync#update change->w-sync#update",
                }
            ),
        ),
    ]
#...
```

The goal here is to ensure that the sync field is basically doing something, firing events and also picking up the right element. Note the usage of `page-is-live` in the selector. This is because of the existing behaviour in `client/src/entrypoints/admin/page-editor.js` where the field should not try to sync with the slug if the page is already live. 

There may be a better way to do this page is live logic, but this should be ok for the initial implementation.

#### Part 3 - Ensure the `w-slug` field can support the new usage

Alongside the existing `w-slug` implementation done in #10086 in `wagtail/admin/panels/page_utils.py` we now want to add support for listening to the `w-sync` events. This is where Stimulus gets pretty powerful.

Our goal is the output in the HTML to be something like this...

```html
<input
  type="text"
  name="slug"
  value="anadama-bread"
  maxlength="255"
  required=""
  id="id_slug"
  data-controller="w-slug"
  data-action="blur->w-slug#slugify w-sync:check->w-slug#compare w-sync:apply->w-slug#urlify:prevent"
  data-w-slug-mode-param="urlify"
/>
```

Essentially we are saying

* If this element gets an event called 'w-sync:check', then call the `compare` method.
* If this element gets an event called 'w-sync:apply', then call the `urlify` method AND prevent default - this will stop the sync behaviour from updating the field's value.
* When any method is dispatched, call it with the params `{ mode }` - (note: we could skip this data attribute and just make the default `urilify` in the destructuring.

We will need to add two new methods to the `SlugController`;

* `slugify` method should also return the value it would have applied
* `compare` - this will receive an event and check the `value` in its detail, then compare the **resolved** version of that value against the current slug field's value. The way it will compare will be via the `mode` (urlify/slugify). If they are the same then call the event's `preventDetault` - also return the boolean value to the method call.
* `urlify` - essentially does the exact same thing as `cleanForSlug(this.value, true)`

### Part 4 - clean up

We can now validate the behaviour, write some upgrade consideration docs and potentially consider documenting this sync field fully.

We should also use the slug field's max length attribute as the max length value that gets passed into URLify.

Additionally, we should split out the `client/src/utils/text.ts` into three files; urlify, slugify and escapeHtml. Where the URLify function is basically the existing vendor script but outside of vendor.

### Describe alternatives you've considered

* Leave as is, but this means we cannot move away from page-editor.js used everywhere.

### Additional context

* This is part of a larger goal to migrate to Stimulus.
* Must be completed after #10086
* This is also part of a larger goal to avoid `wagtail/admin/templates/wagtailadmin/pages/_editor_js.html` being used everywhere and minimise global usage in JS See #2936 & #9771
* This functionality should cover almost all of the requirements in #1063 & #161 - but may need to be solved with example documentation. This sync field approach would not support multiple fields as a source into another field but that can probably be split to a new issue if really needed.

**Comment by Lovelyfin00** [comment]
Working on this

**Comment by Lovelyfin00** [comment]
> ```python
>  "data-action": "focus->w-sync#check blur->w-sync#update change->w-sync#update",
> ```

Is there a mistake here? I thought the action is supposed to be

```
    data-action="focus->w-sync#check change->w-sync#apply cut->w-sync#clear"
````

**Comment by lb-** [comment]
@Lovelyfin00 been thinking that maybe `w-slug` should be renamed to `w-clean` - it seems like it makes sense for this approach to be used for non-slug field cleaning in the future. 

Do you think I should create a new issue for this rename or does it make sense in the scope.of this issue.

**Comment by Lovelyfin00** [comment]
It can be used in other approach I guess. Just that the regex method in the text util might need to be changed based on what it’s been used for. Creating a new issue is ok. Just something simple.

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/issues/10198 has been created

**Comment by Lovelyfin00** [comment]
> * `urlify` - essentially does the exact same thing as `cleanForSlug(this.value, true)`

I don't understand what this method is supposed to do since the `apply` method is already setting the value of the slug input

**Comment by lb-** [comment]
@Lovelyfin00 good question, the idea is that the urlify function will only sit in the slug controller. The sync controller would not do that 'work' itself. 

I'll try to add a comment to your draft PR with a bit more detail, but the goal here is that the sync field tells some other field to update its own value. Think about two fields that are plain text fields you want to keep.in sync. 

However, building on that basic behaviour, we can also let the sync target contol **how** it updates - in the case of the slug field it updates with the urlify'd value.

**Comment by Lovelyfin00** [comment]
Alright
I have just pushed my latest changes

**Comment by lb-** [comment]
Note: I will split out the 'clean up' stuff to a new issue.
